xio-core/src/main/java/com/xjeffrose/xio/http/GrpcRequestHandler.java

```diff
@@ -0,0 +1,283 @@
+        GrpcResponse extends com.google.protobuf.GeneratedMessageV3>
+    implements PipelineRequestHandler {
+
+  private static final String GRPC_TRAILING_HEADER_STATUS_KEY = "grpc-status";
+  private static final String GRPC_CONTENT_TYPE_VALUE = "application/grpc+proto";
+  private static final String GRPC_MESSAGE_NO_COMPRESSION = "compression not supported";
+  private static final String GRPC_MESSAGE_LARGE_SIZE = "payload is too large";
+  private static final String GRPC_MESSAGE_NO_METADATA = "metadata not provided";
+  private static final String GRPC_MESSAGE_WRONG_SIZE =
+      "indicated payload size does not match actual payload size";
+  private static final String GRPC_MESSAGE_CANNOT_MAKE_RESPONSE =
+      "unable to create response object";
+
+  private enum GrpcStatus { // These are defined by gRPC
+    OK("0"),
+    CANCELLED("1"),
+    UNKNOWN("2"),
+    INVALID_ARGUMENT("3"),
+    DEADLINE_EXCEEDED("4"),
+    NOT_FOUND("5"),
+    ALREADY_EXISTS("6"),
+    PERMISSION_DENIED("7"),
+    RESOURCE_EXHAUSTED("8"),
+    FAILED_PRECONDITION("9"),
+    ABORTED("10"),
+    OUT_OF_RANGE("11"),
+    UNIMPLEMENTED("12"),
+    INTERNAL("13"),
+    UNAVAILABLE("14"),
+    DATA_LOSS("15"),
+    UNAUTHENTICATED("16");
+
+    private final String text;
+
+    GrpcStatus(final String text) {
+      this.text = text;
+    }
+
+    @Override
+    public String toString() {
+      return text;
+    }
+  }
+
+  private static class GrpcState {
+    int size;
+    ByteBuf buffer;
+
+    GrpcState(int size, ByteBuf buffer) {
+      this.size = size;
+      this.buffer = buffer;
+    }
+  }
+
+  private static final AttributeKey<HashMap<Integer, GrpcState>> CHANNEL_MESSAGE_SESSION_KEY =
+      AttributeKey.newInstance("xio_grpc_session");
+
+  private static final int sizeTooLargeToHandle = 1_000_000;
+
+  private static HashMap<Integer, GrpcState> lazyCreateSession(ChannelHandlerContext ctx) {
+    HashMap<Integer, GrpcState> session = ctx.channel().attr(CHANNEL_MESSAGE_SESSION_KEY).get();
+    if (session == null) {
+      session = new HashMap<>();
+      ctx.channel().attr(CHANNEL_MESSAGE_SESSION_KEY).set(session);
+    }
+    return session;
+  }
+
+  private final GrpcRequestParser<GrpcRequest> requestParser;
+  private final Function<GrpcRequest, GrpcResponse> appLogic;
+
+  GrpcRequestHandler(
+      GrpcRequestParser<GrpcRequest> requestParser, Function<GrpcRequest, GrpcResponse> appLogic) {
+    this.requestParser = requestParser;
+    this.appLogic = appLogic;
+  }
+
+  @Override
+  public void handle(ChannelHandlerContext ctx, Request request, RouteState route) {
+    ByteBuf actualBuffer;
+    if (request instanceof SegmentedRequestData) {
+      SegmentedRequestData segmentedRequest = (SegmentedRequestData) request;
+      actualBuffer = segmentedRequest.content();
+    } else {
+      actualBuffer = request.body();
+    }
+
+    HashMap<Integer, GrpcState> session = lazyCreateSession(ctx);
```

Comment:
⚪️ If actualBuffer is null or not readable we will still execute these 2 lines...Does that matter?

Reply:
it won't change the logic, but I'll move them down.